In [9]:
import numpy as np
import matplotlib.pyplot as plt
import math
import heapq, sys

# Ejercicio 1

## Diferencia Principal

Dijkstra es un algoritmo que encuentra la ruta más corta entre un nodo de inicio y un nodo de fin. Mientras que Floyd encuentra la distancia más corta entre cada par de vertices.

Para esto es importante la noción de **dirección** y **distancia**, **dirección** es como *avanzamos*, es decir, si de un nodo $A$ existe una ruta directa hacia un nodo $B$ y la **distancia** es cuanto nos cuesta avanzar de un nodo $A$ con una ruta directa a dirección $B$. No necesariamente es el mismo costo ir de $A$ a $B$ que de $B$ a $A$.


# Definicion de Algoritmos.

In [40]:
def dijkstras(adjMatrix, sourceNode, destNode):
    distances = {}
    path = {}
    for node in range(len(adjMatrix)):
        distances[node] = float(sys.maxsize)
        path[node] = None
    distances[sourceNode] = 0
    pq = []
    for node in range(len(adjMatrix)):
        heapq.heappush(pq, (node, distances[node]))

    while len(pq) != 0:
        currNode, currNodeDistance = heapq.heappop(pq)
        for neighborNode in range(len(adjMatrix[currNode])):
            if adjMatrix[currNode][neighborNode] > 0:
                weight = adjMatrix[currNode][neighborNode]
                distance = currNodeDistance + weight
                if distance < distances[neighborNode]:
                    distances[neighborNode] = distance
                    heapq.heappush(pq, (neighborNode, distance))
                    path[neighborNode] = currNode
    return path, distances[destNode]

In [42]:
def floyd(adjMatrix):
    min_paths = []
    for i in range(len(adjMatrix)):
        for j in range(len(adjMatrix)):
            if(i != j):
                path, dist = dijkstras(adjMatrix, i, j)
                min_paths.append({"Start": i+1, "End":j+1, "route": path, "distance":dist })
    return min_paths

In [43]:
def pathString(path, startVertex, endVertex):
    curr = endVertex
    listOfPaths = [curr]
    while(curr != startVertex):
        curr = path[curr]
        listOfPaths[:0] = [curr]
    s = str(listOfPaths[0]+1)
    for i in range(1, len(listOfPaths)):
        if i < len(listOfPaths) - 1:
            s += "-" + str(listOfPaths[i]+1)
        else:
            str(listOfPaths[i]+1)
    return s + "-"+str(endVertex+1)

# Ejercicio 2

Modelamos el sistema como un problema de ruta más corta. Con lo que obtenemos un diagrama como este:

![](diagrama2.png)

Con lo que construimos la matriz de distancia para encontrar la ruta mas corta de $1 \rightarrow 5$

In [51]:
distance_matrix = [
    [0, 1600,3868,7522,10870],
    [-1,   0, 1780,4552,7144],
    [-1,-1,     0, 2200,4216],
    [-1,-1,-1,        0,2620],
    [-1,-1,-1,-1,          0]
]
route, distance = dijkstras(distance_matrix, 0, 4)
print("Distancia Minima de 1 a 5: " + str(distance) + \
      ", Ruta Optima: "+ pathString(route,0,4))

Distancia Minima de 1 a 5: 7596, Ruta Optima: 1-2-3-5


Vemos que la distancia minima es 7596, que representa el costo total y la forma de obtener ese costo minimo es:

- En el mes 1: Ordenar unicamente las unidades para el mes 1, es decir 100
- En el mes 2: Ordenar unicamente las unidades para el mes 2, es decir, 140
- En el mes 3: Ordenar las unidades para el mes 3 y mes 4, es decir, 390
- El ultimo estado (mes 5): es *flotante* solo un punto de meta, no se ordena nada, pero sirve para indicar las ordenes que entran en el mes 4. 

# Ejercicio 3

Aplique  el  algoritmo  de  Floyd  a  la  red  de  la  siguiente  figura.   Los  arcos  (7,6)  y  (6,4)  sonunidireccionales, y todas las distancias est ́an en millas.  Determine la ruta m ́as corta entre lossiguientes pares de nodos

## Floyd

In [39]:
distance_matrix = [
    [0, 5, 3, -1,-1,-1,-1],
    [5, 0, 1, 5, 2, -1,-1],
    [3, 1, 0, 7,-1, -1,12],
    [-1,5, 7, 0, 3, -1, 3],
    [-1,2,-1, 3, 0, 1, -1],
    [-1,-1,-1,1, 1, 0, -1],
    [-1,-1, 12,3,-1, 4, 0]
]

In [47]:
routes = floyd(distance_matrix)
for route in routes:
    print(" Inicio: " + str(route["Start"]) + \
          " Fin: " + str(route["End"]) + \
          ", Distancia Minima: "+ str(route["distance"]) + \
          " Ruta: " + pathString(route["route"], route["Start"]-1, route["End"]-1))

 Inicio: 1 Fin: 2, Distancia Minima: 4 Ruta: 1-3-2
 Inicio: 1 Fin: 3, Distancia Minima: 3 Ruta: 1-3
 Inicio: 1 Fin: 4, Distancia Minima: 8 Ruta: 1-3-2-5-6-4
 Inicio: 1 Fin: 5, Distancia Minima: 6 Ruta: 1-3-2-5
 Inicio: 1 Fin: 6, Distancia Minima: 7 Ruta: 1-3-2-5-6
 Inicio: 1 Fin: 7, Distancia Minima: 11 Ruta: 1-3-2-5-6-4-7
 Inicio: 2 Fin: 1, Distancia Minima: 4 Ruta: 2-3-1
 Inicio: 2 Fin: 3, Distancia Minima: 1 Ruta: 2-3
 Inicio: 2 Fin: 4, Distancia Minima: 4 Ruta: 2-5-6-4
 Inicio: 2 Fin: 5, Distancia Minima: 2 Ruta: 2-5
 Inicio: 2 Fin: 6, Distancia Minima: 3 Ruta: 2-5-6
 Inicio: 2 Fin: 7, Distancia Minima: 7 Ruta: 2-5-6-4-7
 Inicio: 3 Fin: 1, Distancia Minima: 3 Ruta: 3-1
 Inicio: 3 Fin: 2, Distancia Minima: 1 Ruta: 3-2
 Inicio: 3 Fin: 4, Distancia Minima: 5 Ruta: 3-2-5-6-4
 Inicio: 3 Fin: 5, Distancia Minima: 3 Ruta: 3-2-5
 Inicio: 3 Fin: 6, Distancia Minima: 4 Ruta: 3-2-5-6
 Inicio: 3 Fin: 7, Distancia Minima: 8 Ruta: 3-2-5-6-4-7
 Inicio: 4 Fin: 1, Distancia Minima: 9 Ruta: 4-2-3-1


Vemos que la ruta de $1 \rightarrow 7$ es diferente a la de $7 \rightarrow 1$, pues hay trayectorias unidireccionales, es decir, podemos llegar de $A \rightarrow B$ pero no podemos llegar de $A \leftarrow B$.

Ruta de $1 \rightarrow 7$: $1\rightarrow3\rightarrow2\rightarrow5\rightarrow6\rightarrow4\rightarrow7$

Ruta de $7 \rightarrow 1$: $7\rightarrow6\rightarrow5\rightarrow2\rightarrow3\rightarrow1$